## Sorting the new data into spending sectors

In [3]:
import pandas as pd
import numpy as np

In [4]:
transaction_df = pd.read_csv('C:/Users/camer/OneDrive/Documents/Data Science MSc/DSMP/simulated_transaction_2024.csv')
transaction_df

Date Timestamp   Account No       Balance   Amount  \
0       01/01/2023     00:00  678330503.0   2971.000000  1584.00   
1       01/01/2023     00:00  472213568.0   3792.000000  1950.00   
2       01/01/2023     00:00  472213568.0   3012.000000  -780.00   
3       01/01/2023     00:00  283027736.0   1787.000000   780.00   
4       01/01/2023     00:00  624500124.0   3226.000000  1825.00   
...            ...       ...          ...           ...      ...   
230591  06/12/2023     20:54  581655972.0  45935.206861   -41.06   
230592  06/12/2023     20:55  786141370.0   -244.837500   -62.35   
230593  06/12/2023     21:05  824916823.0   9709.172159   -32.94   
230594  06/12/2023     21:13  366550080.0  26834.165794   -19.25   
230595  06/12/2023     21:32  259015404.0   1145.349837   -37.64   

        Third Party Account No    Third Party Name  
0                          NaN  Westport Care Home  
1                          NaN    Barbiee Boutique  
2                  283027736.0                 NaN  
3                  472213568.0                 NaN  
4                          NaN            Fat Face  
...                        ...                 ...  
230591                     NaN               Tesco  
230592                     NaN     Sainsbury Local  
230593                     NaN           Deliveroo  
230594                     NaN              Amazon  
230595                     NaN             JustEat  

[230596 rows x 7 columns]

Selecting only rows with Third party names

In [5]:
business_transactions = transaction_df[~transaction_df['Third Party Name'].isnull()]
business_transactions = business_transactions.drop(['Third Party Account No', 'Balance', 'Timestamp', 'Date'], axis=1)
print(f"{business_transactions['Third Party Name'].value_counts()}")
print(f"\n\n Names of third party accounts: \n{business_transactions['Third Party Name'].unique()}")

Third Party Name
Tesco                 34108
Sports Direct         16259
Topshop               10332
Fat Face               9543
PureGym                8293
                      ...  
Millets                   1
Gap Kids                  1
Specsavers                1
RugbyFields               1
Mountain Warehouse        1
Name: count, Length: 83, dtype: int64


 Names of third party accounts: 
['Westport Care Home' 'Barbiee Boutique' 'Fat Face' 'Lavender Primary'
 'A Cut Above' 'Tesco' 'Brilliant Brushes' 'Costa Coffee'
 'Premier Finance' 'North Face' 'Craftastic' 'Fitted Stitch' 'Selfridges'
 'Starbucks' 'The Crown' 'Cass Art' 'Green Park Academy' 'Rose & Crown'
 'Kings Arms' 'Sunny Care Nursery' 'A Yarn Story' 'Pets Corner'
 'Lavender Fields' 'Victoria Park' 'Halifax' 'LBG' 'Blizzard' 'Xbox'
 'Mojang Studios' 'PureGym' 'Disney' 'Netflix' 'Grand Union BJJ' 'Amazon'
 'SquareOnix' 'Deliveroo' 'JustEat' 'Coop Local' 'Sainsbury' 'AMAZON'
 'The Works' "Blackwell's" 'Topshop' 'Matalan' 'Fo

In [6]:
null_values = business_transactions[business_transactions.isnull().any(axis=1)]
null_values

Account No   Amount Third Party Name
527             NaN -1143.00          Halifax
949     857754342.0      NaN              LBG
1295            NaN   -18.99          PureGym
1335            NaN   -18.99          PureGym
1563            NaN   -15.99          Netflix
...             ...      ...              ...
227463  460463595.0      NaN            Tesco
227645  587225252.0      NaN  Sainsbury Local
229119          NaN  -225.37          Topshop
229970  497772347.0      NaN          Matalan
230360          NaN  -111.36    Fitted Stitch

[420 rows x 3 columns]

The Null values left in the data can be seen above, this is ontl 420 out of 230000 rows so they can be dropped with minimal effect

In [7]:
business_transactions = business_transactions.dropna()
business_transactions

Account No   Amount    Third Party Name
0       678330503.0  1584.00  Westport Care Home
1       472213568.0  1950.00    Barbiee Boutique
4       624500124.0  1825.00            Fat Face
5       203466392.0  2841.66    Lavender Primary
6       768271776.0  1950.00         A Cut Above
...             ...      ...                 ...
230591  581655972.0   -41.06               Tesco
230592  786141370.0   -62.35     Sainsbury Local
230593  824916823.0   -32.94           Deliveroo
230594  366550080.0   -19.25              Amazon
230595  259015404.0   -37.64             JustEat

[223097 rows x 3 columns]

Some of the transactions are positive, meaning people are being paid, this analysis focusing on spending so these transactions need to be removed

In [8]:
business_spending = business_transactions[business_transactions['Amount'] <= 0]
business_spending

Account No  Amount Third Party Name
495     355733816.0 -664.00          Halifax
496     472213568.0 -600.00              LBG
497     624500124.0 -654.00              LBG
498     203466392.0 -672.00          Halifax
499     768271776.0 -632.00              LBG
...             ...     ...              ...
230591  581655972.0  -41.06            Tesco
230592  786141370.0  -62.35  Sainsbury Local
230593  824916823.0  -32.94        Deliveroo
230594  366550080.0  -19.25           Amazon
230595  259015404.0  -37.64          JustEat

[216044 rows x 3 columns]

In [9]:
business_spending['Account No'].unique()

array([3.55733816e+08, 4.72213568e+08, 6.24500124e+08, 2.03466392e+08,
       7.68271776e+08, 5.64744955e+08, 6.75806859e+08, 4.56221621e+08,
       3.50921975e+08, 3.31450202e+08, 4.55831435e+08, 8.32564361e+08,
       1.72084472e+08, 5.50169791e+08, 1.41521177e+08, 4.42070890e+08,
       7.44968736e+08, 8.65748375e+08, 5.22188082e+08, 4.73820682e+08,
       7.86141370e+08, 8.12839908e+08, 5.03575533e+08, 9.79135922e+08,
       1.34461671e+08, 4.30910686e+08, 6.18514666e+08, 3.61201224e+08,
       3.70757362e+08, 6.37433131e+08, 7.54141345e+08, 5.59995067e+08,
       1.19993184e+08, 4.78986718e+08, 2.98000165e+08, 4.16060965e+08,
       6.73025413e+08, 2.41459011e+08, 4.03902116e+08, 6.04288474e+08,
       6.56781737e+08, 8.71866387e+08, 4.25964226e+08, 2.00080192e+08,
       5.19649533e+08, 1.83888402e+08, 6.80773155e+08, 6.41764023e+08,
       2.15426302e+08, 3.86729293e+08, 2.85428103e+08, 9.25037906e+08,
       8.76890798e+08, 8.97714622e+08, 1.20634201e+08, 8.92354433e+08,
      

In [10]:
print(f"\n\n Names of third party accounts: \n{business_spending['Third Party Name'].unique()}")



 Names of third party accounts: 
['Halifax' 'LBG' 'Blizzard' 'Xbox' 'Mojang Studios' 'PureGym' 'Disney'
 'Netflix' 'Grand Union BJJ' 'Amazon' 'SquareOnix' 'Deliveroo' 'JustEat'
 'Coop Local' 'Selfridges' 'Sainsbury' 'AMAZON' 'The Works' "Blackwell's"
 'Fat Face' 'Topshop' 'Matalan' 'Foyles' 'Tesco' 'Wool' 'Hobby Lobby'
 'Revella' 'Sainsbury Local' 'Starbucks' 'Loosely Fitted'
 'Stitch By Stitch' 'Coffee #1' 'Hobbycraft' 'A Yarn Story' 'Craftastic'
 'Kings Arms' 'Costa Coffee' 'The Crown' 'Lloyds Pharmacy' 'Rose & Crown'
 'Fitted Stitch' 'Gamestation' 'CeX' 'Etsy' 'Five Senses Art'
 'Sports Direct' 'Cass Art' 'Brilliant Brushes' 'Boots' 'Reebok'
 'JD Sports' 'Head' 'Frankie & Bennies' 'Gap Kids' 'North Face'
 'Town High' 'Collector Cave' 'Barbiee Boutique' 'RugbyFields'
 'Mamas & Papas' 'Lavender Primary' 'Remedy plus care' 'Specsavers'
 'Kew House' 'HMV' 'Vision Express' 'Millets' 'Pets Corner' 'Mothercare'
 'A Cut Above' 'Happy Days Home' 'Mountain Warehouse' 'Victoria Park'
 'Unive

Listing cateogries to split the third party accounts into

In [11]:
categories = {
    'Gaming': ['Blizzard', 'Xbox', 'Mojang Studios', 'Gamestation', 'SquareOnix'],
    'Streaming Services': ['Netflix', 'Disney'],
    'Misc/General Retail': ['Etsy', 'Amazon', 'AMAZON', 'Selfridges', 'The Works', 'CeX', 'HMV'],
    'Sport/Fitness': ['Sports Direct', 'RugbyFields', 'PureGym', 'Grand Union BJJ', 'Reebok', 'JD Sports', 'Head', 'Mountain Warehouse', 'Millets'],
    'Food Delivery': ['Deliveroo', 'JustEat'],
    'Supermarket': ['Coop Local', 'Tesco', 'Sainsbury', 'Sainsbury Local'],
    'Hospitality': ['Kings Arms', 'The Crown', 'Rose & Crown', 'Frankie & Bennies'],
    'Art/Hobby Supplies': ['Collector Cave', 'Craftastic', 'Cass Art', 'Hobbycraft', 'Wool', 'A Yarn Story', 'Hobby Lobby', 'Five Senses Art', 'Brilliant Brushes'],
    'Bookshop': ["Blackwell's", 'Foyles'],
    'Children': ['Lavender Primary', 'Gap Kids', 'Mamas & Papas', 'Mothercare', 'Kew House'],
    'Fashion': ['Fat Face', 'Topshop', 'Matalan', 'Revella', 'Fitted Stitch', 'Loosely Fitted', 'Stitch By Stitch', 'North Face', 'Barbiee Boutique'],
    'Coffee': ['Costa Coffee', 'Starbucks', 'Coffee #1'],
    'Healthcare': ['University College Hospital','Specsavers' ,'Vision Express' , 'Lloyds Pharmacy', 'Boots', 'Remedy plus care', 'Happy Days Home'],
}

Some of the tird party names do not fit into any categories/ it is not obvious what they represent, 
'Town High'
'Victoria Park'
'A Cut Above'
'Pets Corner'
these all have 2 or less transactions so they can just be removed. Pets corner is insightful but has only 1 transaction so it is not worth creating a whole new category

In [12]:
values_to_drop = ['Town High', 'Victoria Park', 'A Cut Above', 'Pets Corner', 'Halifax', 'LBG']
business_spending = business_spending[~business_spending['Third Party Name'].isin(values_to_drop)]


In [13]:
businesses = business_spending['Third Party Name'].value_counts()
businesses

Third Party Name
Tesco                          33871
Sports Direct                  16237
Topshop                        10311
Fat Face                        9303
PureGym                         8278
                               ...  
Millets                            1
Mothercare                         1
Happy Days Home                    1
Mountain Warehouse                 1
University College Hospital        1
Name: count, Length: 68, dtype: int64

In [15]:
business_spending['category'] = ''
business_spending = business_spending.copy()


for category, keywords in categories.items():
    business_spending.loc[business_spending['Third Party Name'].isin(keywords), 'category'] = category

business_spending

Account No  Amount Third Party Name             category
997     678330503.0  -14.99         Blizzard               Gaming
998     678330503.0   -7.00             Xbox               Gaming
999     678330503.0   -5.99   Mojang Studios               Gaming
1000    678330503.0   -7.00             Xbox               Gaming
1001    678330503.0  -18.99          PureGym        Sport/Fitness
...             ...     ...              ...                  ...
230591  581655972.0  -41.06            Tesco          Supermarket
230592  786141370.0  -62.35  Sainsbury Local          Supermarket
230593  824916823.0  -32.94        Deliveroo        Food Delivery
230594  366550080.0  -19.25           Amazon  Misc/General Retail
230595  259015404.0  -37.64          JustEat        Food Delivery

[210033 rows x 4 columns]

In [16]:
sender_category_spending = business_spending.groupby(['Account No', 'category'])['Amount'].sum().reset_index()

In [17]:
sender_category_spending

Account No             category   Amount
0     101531259.0   Art/Hobby Supplies  -474.69
1     101531259.0              Fashion -2248.00
2     101531259.0        Food Delivery   -87.27
3     101531259.0        Sport/Fitness -2074.88
4     101531259.0   Streaming Services   -95.88
...           ...                  ...      ...
8090  999752672.0           Healthcare   -10.88
8091  999752672.0  Misc/General Retail  -339.56
8092  999752672.0        Sport/Fitness -1787.63
8093  999752672.0   Streaming Services  -191.88
8094  999752672.0          Supermarket -5602.34

[8095 rows x 3 columns]

In [18]:
pivoted_data = sender_category_spending.pivot(index='Account No', columns='category', values='Amount').reset_index()

pivoted_data = pivoted_data.fillna(0)

pivoted_data

categorical_spending_by_account = pivoted_data

Finding the proportion of the total money spent by each account in each category

In [19]:
categories_only = categorical_spending_by_account.drop('Account No', axis=1)
total_spending_per_account = categories_only.sum(axis=1)
categorical_spending_by_account['total_spending'] = total_spending_per_account

In [20]:
categorical_spending_by_account

category   Account No  Art/Hobby Supplies  Bookshop  Children  Coffee  \
0         101531259.0             -474.69      0.00       0.0    0.00   
1         104832000.0                0.00      0.00       0.0    0.00   
2         105375973.0                0.00   -462.18       0.0    0.00   
3         106601471.0                0.00   -199.82       0.0    0.00   
4         108481285.0             -425.87   -124.88       0.0    0.00   
..                ...                 ...       ...       ...     ...   
971       995615876.0                0.00      0.00       0.0    0.00   
972       996042490.0             -883.33    -59.92       0.0    0.00   
973       998390769.0                0.00      0.00       0.0    0.00   
974       998405607.0              -14.97    -32.96       0.0    0.00   
975       999752672.0                0.00      0.00       0.0 -765.05   

category  Fashion  Food Delivery   Gaming  Healthcare  Hospitality  \
0        -2248.00         -87.27     0.00        0.00          0.0   
1            0.00        -211.33 -1165.64       -7.31          0.0   
2        -3453.31        -134.64     0.00      -69.03        -12.9   
3        -3905.01        -398.88  -944.62     -159.38        -23.9   
4        -4469.60         -64.28     0.00     -660.89          0.0   
..            ...            ...      ...         ...          ...   
971      -6035.54        -383.22  -659.64        0.00        -26.0   
972      -2889.24        -293.79 -1140.70     -110.59          0.0   
973      -3091.43        -147.83 -1763.44        0.00          0.0   
974      -3556.00        -161.30  -511.74        0.00          0.0   
975      -4846.16        -189.31     0.00      -10.88          0.0   

category  Misc/General Retail  Sport/Fitness  Streaming Services  Supermarket  \
0                        0.00       -2074.88              -95.88     -2394.86   
1                       -1.06       -1346.97             -191.88     -3619.33   
2                    -2679.37       -2805.18             -191.88      -258.79   
3                     -655.66        -257.02             -191.88     -1039.83   
4                    -2181.77       -2129.02              -87.89     -1528.08   
..                        ...            ...                 ...          ...   
971                      0.00       -1416.73                0.00     -1469.14   
972                   -439.55       -1769.55             -191.88     -2887.46   
973                    -10.66           0.00             -287.76     -3866.23   
974                    -36.95       -1978.78             -191.88     -2431.58   
975                   -339.56       -1787.63             -191.88     -5602.34   

category  total_spending  
0               -7375.58  
1               -6543.52  
2              -10067.28  
3               -7776.00  
4              -11672.28  
..                   ...  
971             -9990.27  
972            -10666.01  
973             -9167.35  
974             -8916.16  
975            -13732.81  

[976 rows x 15 columns]

Finding the proportion of spending for each account in each category (spending in category/ total spending)

In [21]:
proportion_category_spending = categorical_spending_by_account.iloc[:, 1:-1].div(categorical_spending_by_account['total_spending'], axis=0)

In [22]:
proportion_category_spending

category  Art/Hobby Supplies  Bookshop  Children   Coffee   Fashion  \
0                   0.064360 -0.000000      -0.0 -0.00000  0.304790   
1                  -0.000000 -0.000000      -0.0 -0.00000 -0.000000   
2                  -0.000000  0.045909      -0.0 -0.00000  0.343023   
3                  -0.000000  0.025697      -0.0 -0.00000  0.502188   
4                   0.036486  0.010699      -0.0 -0.00000  0.382924   
..                       ...       ...       ...      ...       ...   
971                -0.000000 -0.000000      -0.0 -0.00000  0.604142   
972                 0.082817  0.005618      -0.0 -0.00000  0.270883   
973                -0.000000 -0.000000      -0.0 -0.00000  0.337222   
974                 0.001679  0.003697      -0.0 -0.00000  0.398826   
975                -0.000000 -0.000000      -0.0  0.05571  0.352889   

category  Food Delivery    Gaming  Healthcare  Hospitality  \
0              0.011832 -0.000000   -0.000000    -0.000000   
1              0.032296  0.178137    0.001117    -0.000000   
2              0.013374 -0.000000    0.006857     0.001281   
3              0.051296  0.121479    0.020496     0.003074   
4              0.005507 -0.000000    0.056620    -0.000000   
..                  ...       ...         ...          ...   
971            0.038359  0.066028   -0.000000     0.002603   
972            0.027545  0.106947    0.010368    -0.000000   
973            0.016126  0.192361   -0.000000    -0.000000   
974            0.018091  0.057395   -0.000000    -0.000000   
975            0.013785 -0.000000    0.000792    -0.000000   

category  Misc/General Retail  Sport/Fitness  Streaming Services  Supermarket  
0                   -0.000000       0.281318            0.013000     0.324701  
1                    0.000162       0.205848            0.029324     0.553117  
2                    0.266146       0.278643            0.019060     0.025706  
3                    0.084318       0.033053            0.024676     0.133723  
4                    0.186919       0.182400            0.007530     0.130915  
..                        ...            ...                 ...          ...  
971                 -0.000000       0.141811           -0.000000     0.147057  
972                  0.041210       0.165906            0.017990     0.270716  
973                  0.001163      -0.000000            0.031390     0.421739  
974                  0.004144       0.221932            0.021520     0.272716  
975                  0.024726       0.130172            0.013972     0.407953  

[976 rows x 13 columns]

Finding the average spending proportions per category and then calculating the over/under spending of each account in each category. This means the categories will be standardised (if average spending in one sector is 35%, it may dominate a model and therefore reduce insights into smaller categories)

In [23]:
average_spending_per_category = proportion_category_spending.mean()
average_spending_per_category

category
Art/Hobby Supplies     0.032050
Bookshop               0.008692
Children               0.000271
Coffee                 0.003773
Fashion                0.305412
Food Delivery          0.034542
Gaming                 0.048969
Healthcare             0.006649
Hospitality            0.000859
Misc/General Retail    0.049264
Sport/Fitness          0.169660
Streaming Services     0.021299
Supermarket            0.318558
dtype: float64

In [24]:
over_under_spending = proportion_category_spending.sub(average_spending_per_category, axis=1)
over_under_spending.insert(0, 'Account No', pivoted_data['Account No'])
over_under_spending

category   Account No  Art/Hobby Supplies  Bookshop  Children    Coffee  \
0         101531259.0            0.032309 -0.008692 -0.000271 -0.003773   
1         104832000.0           -0.032050 -0.008692 -0.000271 -0.003773   
2         105375973.0           -0.032050  0.037217 -0.000271 -0.003773   
3         106601471.0           -0.032050  0.017005 -0.000271 -0.003773   
4         108481285.0            0.004435  0.002007 -0.000271 -0.003773   
..                ...                 ...       ...       ...       ...   
971       995615876.0           -0.032050 -0.008692 -0.000271 -0.003773   
972       996042490.0            0.050767 -0.003074 -0.000271 -0.003773   
973       998390769.0           -0.032050 -0.008692 -0.000271 -0.003773   
974       998405607.0           -0.030371 -0.004995 -0.000271 -0.003773   
975       999752672.0           -0.032050 -0.008692 -0.000271  0.051936   

category   Fashion  Food Delivery    Gaming  Healthcare  Hospitality  \
0        -0.000623      -0.022710 -0.048969   -0.006649    -0.000859   
1        -0.305412      -0.002246  0.129168   -0.005532    -0.000859   
2         0.037611      -0.021168 -0.048969    0.000207     0.000422   
3         0.196775       0.016754  0.072510    0.013847     0.002214   
4         0.077512      -0.029035 -0.048969    0.049971    -0.000859   
..             ...            ...       ...         ...          ...   
971       0.298729       0.003817  0.017060   -0.006649     0.001743   
972      -0.034529      -0.006998  0.057979    0.003719    -0.000859   
973       0.031809      -0.018416  0.143392   -0.006649    -0.000859   
974       0.093414      -0.016451  0.008426   -0.006649    -0.000859   
975       0.047477      -0.020757 -0.048969   -0.005857    -0.000859   

category  Misc/General Retail  Sport/Fitness  Streaming Services  Supermarket  
0                   -0.049264       0.111657           -0.008300     0.006143  
1                   -0.049102       0.036188            0.008024     0.234558  
2                    0.216883       0.108983           -0.002240    -0.292852  
3                    0.035055      -0.136607            0.003377    -0.184835  
4                    0.137655       0.012739           -0.013769    -0.187643  
..                        ...            ...                 ...          ...  
971                 -0.049264      -0.027849           -0.021299    -0.171501  
972                 -0.008053      -0.003755           -0.003309    -0.047842  
973                 -0.048101      -0.169660            0.010090     0.103181  
974                 -0.045120       0.052272            0.000221    -0.045842  
975                 -0.024538      -0.039488           -0.007327     0.089395  

[976 rows x 14 columns]

In [27]:
over_under_spending.to_csv('C:/Users/camer/OneDrive/Documents/Data Science MSc/DSMP/New_data_proportional_spending_NO_FS.csv', index=False)